artifact/v3/header/header.cpp

```diff
@@ -0,0 +1,155 @@
+			parser_error::Code::ParseError,
+			"Got unexpected token : '" + tok.TypeToString() + "' expected 'header-info'"));
+	}
+
+	auto expected_info = header::info::Parse(*tok.value);
+
+	if (!expected_info) {
+		return expected::unexpected(parser_error::MakeError(
+			parser_error::Code::ParseError,
+			"Failed to parse the header-info: " + expected_info.error().message));
+	}
+	header.info = expected_info.value();
+
+	tok = lexer.Next();
+	vector<ArtifactScript> state_scripts {};
+	while (tok.type == token::Type::ArtifactScripts) {
+		log::Trace("Parsing state script...");
+		const string artifact_script_path =
+			conf.artifact_scripts_filesystem_path + "/" + tok.TypeToString();
+		ofstream myfile(artifact_script_path);
+		log::Trace("state script name: " + tok.TypeToString());
+		if (!myfile.good()) {
+			// TODO - Return what sort of error (?)
+			return expected::unexpected(parser_error::MakeError(
+				parser_error::Code::ParseError,
+				"Failed to create a file for writing the Artifact script: " + artifact_script_path
+					+ " to the filesystem"));
+		}
+		io::StreamWriter bw {myfile};
+
+		auto err = io::Copy(bw, *tok.value);
+		if (err) {
+			return expected::unexpected(err);
+		}
+
+		state_scripts.push_back(artifact_script_path);
+
+		tok = lexer.Next();
+	}
+
+	header.artifactScripts = state_scripts;
+
+	vector<SubHeader> subheaders {};
+
+	size_t i {0};
+	while (tok.type != token::Type::EOFToken) {
+		log::Trace("Parsing the sub-header ...");
+		SubHeader sub_header {};
+		if (tok.type != token::Type::ArtifactHeaderTypeInfo) {
+			return expected::unexpected(parser_error::MakeError(
+				parser_error::Code::ParseError,
+				"Unexpected entry: " + tok.TypeToString() + " expected: type-info"));
+		}
+		auto expected_type_info = type_info::Parse(*tok.value);
+		if (!expected_type_info) {
+			return expected::unexpected(expected_type_info.error());
+		}
+		sub_header.type_info = expected_type_info.value();
+
+		// NOTE (workaround): Bug in the Artifact format writer:
+		// If the type is a RootfsImage, then the payload-type will be empty. This
+		// is a bug in the mender-artifact tool, which writes the payload. For now,
+		// just work around it.
+		if (header.info.payloads[i].first == Payload::RootfsImage) {
+			log::Debug("Setting the type-info in payload nr " + to_string(i) + " to rootfs-image");
+			sub_header.type_info.type = "rootfs-image";
+		}
+
+		tok = lexer.Next();
+
+		log::Trace("sub-header: looking for meta-data");
+
+		// meta-data (optional)
+		if (tok.type == token::Type::ArtifactHeaderMetaData) {
+			// auto expected_meta_data = meta_data::Parse(*tok.value);
+			// if (!expected_meta_data) {
+			// TODO - follow-up (MEN-6204)
+			// return expected::unexpected(expected_meta_data.error());
+			// sub_header.metadata = expected_meta_data.value();
+			// }
+		}
+		log::Trace("sub-header: parsed the meta-data");
+
+		header.subHeaders.push_back(sub_header);
+
+		tok = lexer.Next();
```

Comment:
Shouldn't this `Next()` be inside the above `if`? Otherwise this will skip a token if it's not `ArtifactHeaderMetaData`, right?

Reply:
Absolutely right!!!

Good :eyes: